### PR TITLE
(PUP-7678) Add functions any and all

### PR DESCRIPTION
**Pull request by hlindberg**
This adds the functions `any()` and `all()` (with functionality comparable to Ruby's `any?` and `all?`.

### CODE REVIEW

File: lib/puppet/functions/all.rb [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    block_param 'Callable[1,1]', :block
+  end
+
+  def all_Hash_1(hash)
+    enumerator = hash.each_pair
+    hash.size.times do
+      return false unless yield(enumerator.next)
+    end
+    true
+  end
```
Comment (thallgren) [inline lib/puppet/functions/all.rb]: Why doesn't this simply do:
```ruby
hash.each_pair { |x| return false unless yield(x) }
return true
```
That will work with using a block with both one and two arguments so you would probably not need both all_Hash_1 and all_Hash_2,

File: lib/puppet/functions/all.rb [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+  def all_Enumerable_1(enumerable)
+    enum = Puppet::Pops::Types::Iterable.asserted_iterable(self, enumerable)
+      begin
+        loop { return false unless yield(enum.next) }
+      rescue StopIteration
+      end
+    true
+  end
```
Comment (thallgren) [inline lib/puppet/functions/all.rb]: Looks like this method can be simplified into:
```ruby
Puppet::Pops::Types::Iterable.asserted_iterable(self, enumerable).all? { |e| yield(e) }
```

File: lib/puppet/functions/all.rb [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    index = 0
+    begin
+      loop do
+        return false unless yield(index, enum.next)
+        index += 1
+      end
+    rescue StopIteration
+    end
+    return true
+  end
```
Comment (thallgren) [inline lib/puppet/functions/all.rb]: How about:
```ruby
enum.each_with_index { |e, i| return false unless(i, e) }
return true
```

File: lib/puppet/functions/all.rb (hunk 1 above)
Comment (hlindberg) [inline lib/puppet/functions/all.rb]: using 'all?' is simpler - changing to using that. I kept 1 or 2 lambda args check in the dispatch - the alternative would be to reflect the given block to see if it has one or two parameters - found no other simple solution.

File: lib/puppet/functions/all.rb (hunk 3 above)
Comment (hlindberg) [inline lib/puppet/functions/all.rb]: Works better with a `yield` also thrown into the mix ;-)
I change this.

**Pull request merged**